A mish-mash of micro-optimizations

nnethercote opened the pull request:
These were aimed at speeding up LLVM codegen, but ended up affecting other places as well.

r? @bjorn3

nnethercote committed:
Avoid unnecessary line lookup.

`lookup_debug_loc` calls `SourceMap::lookup_line`, which does a binary
search over the files, and then a binary search over the lines within
the found file. It then calls `SourceFile::line_begin_pos`, which redoes
the binary search over the lines within the found file.

This commit removes the second binary search over the lines, instead
getting the line starting pos directly using the result of the first
binary search over the lines.

(And likewise for `get_span_loc`, in the cranelift backend.)

nnethercote committed:
Replace a `lookup_debug_loc` call.

`lookup_debug_loc` finds a file, line, and column, which requires two
binary searches. But this call site only needs the file.

This commit replaces the call with `lookup_source_file`, which does a
single binary search.

nnethercote committed:
Use `partition_point` in `SourceMap::lookup_source_file_idx`.

This makes it (a) a little simpler, and (b) more similar to
`SourceFile::lookup_line`.

nnethercote committed:
Avoid an unnecessary use of `SmallStr`.

I don't know why `SmallStr` was used here; some ad hoc profiling showed
this code is not that hot, the string is usually empty, and when it's
not empty it's usually very short. However, the use of a
`SmallStr<1024>` does result in 1024 byte `memcpy` call on each
execution, which shows up when I do `memcpy` profiling. So using a
normal string makes the code both simpler and very slightly faster.

nnethercote committed:
Set capacity of the string passed to `push_item_name`.

Other callsites already do this, but these two were missed. This avoids
some allocations.

nnethercote committed:
Use `SmallVec` for the `bundles` vectors.

They never have a length of more than two. So this commit changes them
to `SmallVec<[_; 2]>`.

Also, we possibly push `None` values and then filter those `None` values
out again with `retain`. So this commit removes the `retain` and instead
only pushes the values if they are `Some(_)`.

nnethercote committed:
Simplify the `bundles` vectors.

After the last commit, they contain `Option<&OperandBundleDef<'a>>` but
the values are always `Some(_)`. This commit removes the needless
`Option` wrapper. This also simplifies the type signatures of
`LLVMRustBuild{Invoke,Call}`, which were relying on the fact that the
represention of `Option<&T>` is the same as `&T` for non-`None` values.

nnethercote committed:
Avoid cloning `LocalDecls`.

`DerefChecker` can just hold a reference instead. This avoids quite a
lot of allocations for some benchmarks.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5398437307/jobs/9805806647)
Build commit: aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513 (`aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5398437307/jobs/9805806647"},"merge_sha":"aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5398437307/jobs/9805806647)
Build commit: aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513 (`aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5398437307/jobs/9805806647"},"merge_sha":"aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513"} -->

oli-obk commented:
@rust-timer build aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513

rust-timer commented:
Queued aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513 with parent 08fd6f719ee764cf62659ddf481e22dbfe4b8894, future [comparison URL](https://perf.rust-lang.org/compare.html?start=08fd6f719ee764cf62659ddf481e22dbfe4b8894&end=aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513): [comparison URL](https://perf.rust-lang.org/compare.html?start=08fd6f719ee764cf62659ddf481e22dbfe4b8894&end=aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=08fd6f719ee764cf62659ddf481e22dbfe4b8894&end=aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.8%, -0.4%] | 11    |
| Improvements ✅ <br /> (secondary) | -1.4% | [-2.0%, -0.9%] | 19    |
| All ❌✅ (primary)                 | -0.9% | [-1.8%, -0.4%] | 11    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=08fd6f719ee764cf62659ddf481e22dbfe4b8894&end=aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.2%  | [2.2%, 5.0%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-1.6%, -1.6%] | 1     |
| All ❌✅ (primary)                 | 3.2%  | [2.2%, 5.0%]   | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=08fd6f719ee764cf62659ddf481e22dbfe4b8894&end=aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.9%, 1.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-2.9%, -1.4%] | 5     |
| Improvements ✅ <br /> (secondary) | -5.6% | [-8.8%, -2.8%] | 11    |
| All ❌✅ (primary)                 | -1.9% | [-2.9%, -1.4%] | 5     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=08fd6f719ee764cf62659ddf481e22dbfe4b8894&end=aa657f7fbe68ce69f72c77fb7c6b8d0c12aa4513&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 662.942s -> 662.822s (-0.02%)

nnethercote committed:
Remove `SmallStr`.

It no longer has any uses. If it's needed in the future, it can be
easily reinstated. Or a crate such as `smallstr` can be used, much like
we use `smallvec`.

nnethercote commented:
The walltime/cycles improvements of up to 8% for `tt-muncher` look spurious; I think it has just become noisy on those metrics. Everything else should be real, though. I think the last commit ("Avoid cloning `LocalDecls`) is the most effective change.

oli-obk commented on compiler/rustc_codegen_llvm/src/builder.rs:
```diff
@@ -225,17 +226,21 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
-        let mut bundles = vec![funclet_bundle];
+        let mut bundles: SmallVec<[_; 2]> = SmallVec::new();
+        if funclet_bundle.is_some() {
+            bundles.push(funclet_bundle);
```
maybe we could stop putting options in there and just use the value directly in LLVMRustBuildCall?

oli-obk replied:
🤣 I should have looked at the next commit

oli-obk commented:
r? @oli-obk 

@bors r+

bors commented:
:pushpin: Commit 7e786e81b00cf48a664084d30d4f82f408825397 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 7e786e81b00cf48a664084d30d4f82f408825397 -->
<!-- homu: {"type":"Approved","sha":"7e786e81b00cf48a664084d30d4f82f408825397","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7e786e81b00cf48a664084d30d4f82f408825397 with merge 8aed93d912ec23819c08e9a89ca1fb461b3cd2e6...
<!-- homu: {"type":"BuildStarted","head_sha":"7e786e81b00cf48a664084d30d4f82f408825397","merge_sha":"8aed93d912ec23819c08e9a89ca1fb461b3cd2e6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5418042043/jobs/9851318930)
Approved by: oli-obk
Pushing 8aed93d912ec23819c08e9a89ca1fb461b3cd2e6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5418042043/jobs/9851318930"},"merge_sha":"8aed93d912ec23819c08e9a89ca1fb461b3cd2e6"} -->

rust-timer commented:
Finished benchmarking commit (8aed93d912ec23819c08e9a89ca1fb461b3cd2e6): [comparison URL](https://perf.rust-lang.org/compare.html?start=330727467b8fdf2c43b95095a0efae7012c4f83b&end=8aed93d912ec23819c08e9a89ca1fb461b3cd2e6&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=330727467b8fdf2c43b95095a0efae7012c4f83b&end=8aed93d912ec23819c08e9a89ca1fb461b3cd2e6&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-1.7%, -0.3%] | 16    |
| Improvements ✅ <br /> (secondary) | -1.4% | [-2.3%, -0.8%] | 15    |
| All ❌✅ (primary)                 | -0.8% | [-1.7%, -0.3%] | 16    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=330727467b8fdf2c43b95095a0efae7012c4f83b&end=8aed93d912ec23819c08e9a89ca1fb461b3cd2e6&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6%  | [0.4%, 1.0%]   | 4     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-2.0%, -2.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-2.0%, 1.0%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=330727467b8fdf2c43b95095a0efae7012c4f83b&end=8aed93d912ec23819c08e9a89ca1fb461b3cd2e6&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5% | [0.5%, 0.5%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.5% | [0.5%, 0.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=330727467b8fdf2c43b95095a0efae7012c4f83b&end=8aed93d912ec23819c08e9a89ca1fb461b3cd2e6&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 662.146s -> 661.088s (-0.16%)
